src/portal-impl.c

```diff
@@ -548,25 +548,107 @@ warn_please_use_portals_conf (void)
+static PortalImplementation *
+find_fallback_portal_implementation (const char *interface)
 {
-  const char **desktops;
-  GList *l;
-  int i;
+  /* x-d-p-gtk has historically been the portal UI backend used by desktop
+   * environments with no backend of their own.
+   * If it isn't installed, that is not an error: we just don't use it. */
+  for (const GList *l = implementations; l != NULL; l = l->next)
+    {
+      PortalImplementation *impl = l->data;
 
-  for (l = implementations; l != NULL; l = l->next)
+      if (!g_str_equal (impl->dbus_name, "org.freedesktop.impl.portal.desktop.gtk"))
+        continue;
+
+      if (!g_strv_contains ((const char **)impl->interfaces, interface))
+        continue;
+
+      g_warning ("Choosing %s.portal for %s as a last-resort fallback",
+                 impl->source, interface);
+      warn_please_use_portals_conf ();
+      return impl;
+    }
+
+  return NULL;
+}
+
+static PortalImplementation *
+find_any_portal_implementation (const char *interface)
+{
+  for (const GList *l = implementations; l != NULL; l = l->next)
     {
       PortalImplementation *impl = l->data;
 
       if (!g_strv_contains ((const char **)impl->interfaces, interface))
         continue;
 
-      if (portal_impl_matches_config (impl, interface))
+      g_debug ("Falling back to %s.portal for %s", impl->source, interface);
+      return impl;
+    }
+
+  return find_fallback_portal_implementation (interface);
+}
+
+static PortalImplementation *
+find_portal_implementation_iface (const PortalInterface *iface)
+{
+  for (char *const *portal = iface->portals; *portal; portal++)
+    {
+      g_debug ("Found '%s' in configuration for %s", *portal, iface->dbus_name);
+
+      if (g_strcmp0 (*portal, "none") == 0)
+        return NULL;
+
+      if (g_strcmp0 (*portal, "*") == 0)
+        return find_any_portal_implementation (iface->dbus_name);
+
+      for (const GList *l = implementations; l != NULL; l = l->next)
         {
-          g_debug ("Using %s.portal for %s (config)", impl->source, interface);
-          return impl;
+          PortalImplementation *impl = l->data;
+
+          if (g_strcmp0 (*portal, impl->source) == 0)
+            return impl;
         }
+
+      g_debug ("%s.portal is unrecognized", *portal);
+    }
+
+  return find_any_portal_implementation (iface->dbus_name);
+}
+
+PortalImplementation *
+find_portal_implementation (const char *interface)
+{
+  PortalImplementation *impl;
+  const char **desktops;
+  GList *l;
+  int i;
+
+  if (config)
+    {
+      /* Interfaces have precedence over the "default" catch all,
+       * to allow for specific interfaces to override the default
+       */
+      for (i = 0; i < config->n_ifaces; i++)
+        {
+          const PortalInterface *iface = config->interfaces[i];
+
+          if (g_strcmp0 (iface->dbus_name, interface) == 0)
+	    {
+	      impl = find_portal_implementation_iface (iface);
+	      goto skip_default;
```

Comment:
Hm, I'd rather avoid this `goto`. I think this whole block can be written as:

```c
  if (config)
    {
      PortalImplementation *impl = NULL;

      /* Interfaces have precedence over the "default" catch all,
       * to allow for specific interfaces to override the default
       */
      for (i = 0; i < config->n_ifaces; i++)
        {
          const PortalInterface *iface = config->interfaces[i];

          if (g_strcmp0 (iface->dbus_name, interface) == 0)
            {
              impl = find_portal_implementation_iface (iface);
              break;
            }
        }

      if (impl == NULL)
        impl = find_portal_implementation_iface (config->default_portal);

      if (impl != NULL)
        {
          g_debug ("Using %s.portal for %s (config)", impl->source, interface);
          return impl;
        }
    }
```

Reply:
That wouldn't have the same behaviour.  If somebody sets a specific interface to none, your code would have it fall back to the default implementation, which isn't what the configuration asked for.  There are two different null-ish states here — a portal can be specifically disabled, or it can not be explicitly configured, and both cases need to be handled differently.

Reply:
Use a `gboolean` variable then

Reply:
Done.